engine/src/main/java/org/hibernate/validator/internal/engine/path/NodeImpl.java

```diff
@@ -69,6 +68,7 @@
 	private final Class<?> containerClass;
 	private final Integer typeArgumentIndex;
 
+	private int hashCode = -1;
```

Comment:
I'm slightly concerned about this change, doesn't it potentially impact safe publication to other threads (it's part of of the published `ConstraintViolation`)?

Reply:
I understand your concern but it's already done like that in `PathImpl` for a long time. So, either we need to get rid of that optimization in both or we decide it's safe in both.

Reply:
I'd prefer to stay on the safer side in both cases

Reply:
Does it matter that we impact safe publication here? The failure mode is to re-calculate the hash in a separate thread, so at worst we're duplicating some effort. 

The String class in the JDK does the same optimization.

Reply:
Yes, that's good reasoning.